src/databricks/sql/cloudfetch/download_manager.py

```diff
@@ -0,0 +1,139 @@
+        self.cloud_fetch_index = next_row_index
+
+    def get_next_downloaded_file(self, next_row_index):
+        if not self.download_handlers:
+            return None
+
+        # Remove handlers we don't need anymore
+        self._remove_past_handlers(next_row_index)
+
+        # Schedule the downloads
+        self._schedule_downloads()
+
+        # Find next file
+        idx = self._find_next_file_index(next_row_index)
+        if idx is None:
+            return None
+        handler = self.download_handlers[idx]
+
+        # Check (and wait) for download status
+        if self._check_if_download_successful(handler):
+            # Buffer should be empty so set buffer to new ArrowQueue with result_file
+            result = DownloadedFile(
+                handler.result_file,
+                handler.result_link.startRowOffset,
+                handler.result_link.rowCount,
+            )
+            self.cloud_fetch_index += handler.result_link.rowCount
+            self.download_handlers.pop(idx)
+            # Return True upon successful download to continue loop and not force a retry
+            return result
+        # Download was not successful for next download item, force a retry
+        return None
+
+    def _remove_past_handlers(self, next_row_index):
+        """
+        Remove any download handlers whose start to end range doesn't include the next row to be fetched
+        i.e. no need to download
+        """
+        i = 0
+        while i < len(self.download_handlers):
+            result_link = self.download_handlers[i].result_link
+            if result_link.startRowOffset + result_link.rowCount > next_row_index:
+                i += 1
+                continue
+            self.download_handlers.pop(i)
+
+    def _schedule_downloads(self):
+        """
+        Schedule downloads for all download handlers if not already scheduled
+        """
+        for handler in self.download_handlers:
+            if handler.is_download_scheduled:
+                continue
+            try:
+                self.thread_pool.submit(handler.run)
+            except:
+                break
+            handler.is_download_scheduled = True
+
+    def _find_next_file_index(self, next_row_index):
+        # Get the next downloaded file
+        next_indices = [i for i, handler in enumerate(self.download_handlers)
+                        if handler.is_download_scheduled and handler.result_link.startRowOffset == next_row_index]
+        return next_indices[0] if len(next_indices) > 0 else None
+
+    def _check_if_download_successful(self, handler):
+        if not handler.is_file_download_successful():
+            if handler.is_link_expired:
+                self._stop_all_downloads_and_clear_handlers()
+                self.fetch_need_retry = True
+                return False
+            elif handler.is_download_timedout:
+                if self.num_consecutive_result_file_download_retries >= \
```

Comment:
I believe either `requests` or `urllib3` (which `requests` uses under the hood) can automatically handle retries for basic HTTP requests like this. We don't need to implement it ourselves.

Reply:
Gotcha. For now, I'll keep this but will explore your suggestion for a future performance improvement